src/core/lib/security/credentials/external/aws_request_signer.cc

```diff
@@ -0,0 +1,207 @@
+//
+// Copyright 2020 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#include <grpc/support/port_platform.h>
+
+#include "src/core/lib/security/credentials/external/aws_request_signer.h"
+
+#include "src/core/lib/uri/uri_parser.h"
+
+#include "absl/strings/ascii.h"
+#include "absl/strings/escaping.h"
+#include "absl/strings/str_format.h"
+#include "absl/strings/str_join.h"
+#include "absl/strings/str_split.h"
+#include "absl/time/clock.h"
+#include "absl/time/time.h"
+
+#include <openssl/hmac.h>
+#include <openssl/sha.h>
+
+namespace grpc_core {
+
+namespace {
+
+const char kAlgorithm[] = "AWS4-HMAC-SHA256";
+const char kDateFormat[] = "%a, %d %b %E4Y %H:%M:%S %Z";
+const char kXAmzDateFormat[] = "%Y%m%dT%H%M%SZ";
+
+void SHA256(const std::string& str, unsigned char out[SHA256_DIGEST_LENGTH]) {
+  SHA256_CTX sha256;
+  SHA256_Init(&sha256);
+  SHA256_Update(&sha256, str.c_str(), str.size());
+  SHA256_Final(out, &sha256);
+}
+
+std::string SHA256Hex(const std::string& str) {
+  unsigned char hash[SHA256_DIGEST_LENGTH];
+  SHA256(str, hash);
+  std::string hash_str(reinterpret_cast<char const*>(hash),
+                       SHA256_DIGEST_LENGTH);
+  return absl::BytesToHexString(hash_str);
+}
+
+std::string HMAC(const std::string& key, const std::string& msg) {
+  unsigned int len;
+  unsigned char digest[EVP_MAX_MD_SIZE];
+  HMAC(EVP_sha256(), key.c_str(), key.length(),
+       (const unsigned char*)msg.c_str(), msg.length(), digest, &len);
+  return std::string(digest, digest + len);
+}
```

Comment:
Hey @jiangtaoli2016, there're some cases that we need the functionalities of `SHA256` and `HMAC`. Currently I directly depend on the openssl. Do you know what's the best way to get these functionalities in gRPC? Thanks!

Reply:
`HMAC()` and `EVP_sha_256()` are both generic crypto functions supported by both OpenSSL and BoringSSL. I don't see a strong reason to wrap under TSI.

Current grpc_secure depends on tsi, which depends on libssl. This looks fine to me.